server/schedulers/hot_region.go

```diff
@@ -1012,94 +1018,124 @@ func (bs *balanceSolver) isReadyToBuild() bool {
 		bs.cur.region != nil && bs.cur.region.GetID() == bs.cur.srcPeerStat.ID()
 }
 
-func (bs *balanceSolver) buildOperator() (op *operator.Operator, infl *statistics.Influence) {
+func (bs *balanceSolver) buildOperators() (ops []*operator.Operator, infl *statistics.Influence) {
 	if !bs.isReadyToBuild() {
 		return nil, nil
 	}
-	var (
-		err         error
-		typ         string
-		sourceLabel string
-		targetLabel string
-	)
 
 	srcStoreID := bs.cur.srcStore.GetID()
 	dstStoreID := bs.cur.dstStore.GetID()
-	switch bs.opTy {
-	case movePeer:
-		srcPeer := bs.cur.region.GetStorePeer(srcStoreID) // checked in getRegionAndSrcPeer
-		dstPeer := &metapb.Peer{StoreId: dstStoreID, Role: srcPeer.Role}
-		sourceLabel = strconv.FormatUint(srcStoreID, 10)
-		targetLabel = strconv.FormatUint(dstPeer.GetStoreId(), 10)
+	sourceLabel := strconv.FormatUint(srcStoreID, 10)
+	targetLabel := strconv.FormatUint(dstStoreID, 10)
+	dim := ""
+	switch bs.cur.progressiveRank {
+	case -3:
+		dim = "all"
+	case -2:
+		dim = dimToString(bs.secondPriority)
+	case -1:
+		dim = dimToString(bs.firstPriority)
+	}
+
+	var createOperator func(region *core.RegionInfo, srcStoreID, dstStoreID uint64) (op *operator.Operator, typ string, err error)
+	switch bs.rwTy {
+	case statistics.Read:
+		createOperator = bs.createReadOperator
+	case statistics.Write:
+		createOperator = bs.createWriteOperator
+	}
+
+	currentOp, typ, err := createOperator(bs.cur.region, srcStoreID, dstStoreID)
+	if err == nil {
+		bs.decorateOperator(currentOp, sourceLabel, targetLabel, typ, dim)
+		ops = []*operator.Operator{currentOp}
+		infl = &statistics.Influence{
+			Loads: append(bs.cur.srcPeerStat.Loads[:0:0], bs.cur.srcPeerStat.Loads...),
+			Count: 1,
+		}
+	}
+
+	if err != nil {
+		log.Debug("fail to create operator", zap.Stringer("rw-type", bs.rwTy), zap.Stringer("op-type", bs.opTy), errs.ZapError(err))
+		schedulerCounter.WithLabelValues(bs.sche.GetName(), "create-operator-fail").Inc()
+		return nil, nil
+	}
 
-		if bs.rwTy == statistics.Read && bs.cur.region.GetLeader().StoreId == srcStoreID { // move read leader
+	return
+}
+
+func (bs *balanceSolver) createReadOperator(region *core.RegionInfo, srcStoreID, dstStoreID uint64) (op *operator.Operator, typ string, err error) {
+	if region.GetStorePeer(dstStoreID) != nil {
```

Comment:
The old way use `GetStoreVoter`?

Reply:
There are mixed cases in the original code, which are unified into `GetStorePeer`. Considering that `dstStore` is a `Learner`, neither `transfer-leader` nor `move-peer` can be successfully created, so there is no difference.